### ISSUE TITLE
Process stuck in a sleep 

### ISSUE THREAD

**Issue by gmontard** [YusukeIwaki/playwright-ruby-client#264]
### Step To Reproduce / Observed behavior

I'm running some scrapper as part of Sidekiq jobs, but unfortunately sometimes they get stuck forever. 

By doing a dump of Sidekiq I can clearly see that there is an infinite sleep happening under the hood when using Playright gem, here is the trace below:

```bash
2023-09-14T09:48:21.654990+00:00 app[worker.1]: pid=2 tid=3nq WARN: /app/vendor/bundle/ruby/3.2.0/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/promises.rb:775:in `sleep'
2023-09-14T09:48:21.654991+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/promises.rb:775:in `wait'
2023-09-14T09:48:21.654991+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/promises.rb:775:in `block in wait_until_resolved'
2023-09-14T09:48:21.654991+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/promises.rb:771:in `synchronize'
2023-09-14T09:48:21.654991+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/promises.rb:771:in `wait_until_resolved'
2023-09-14T09:48:21.654991+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/promises.rb:1256:in `wait_until_resolved!'
2023-09-14T09:48:21.654991+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/concurrent-ruby-1.1.10/lib/concurrent-ruby/concurrent/promises.rb:1471:in `value!'
2023-09-14T09:48:21.654992+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/playwright-ruby-client-1.37.1/lib/playwright/connection.rb:104:in `send_message_to_server'
2023-09-14T09:48:21.654992+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/playwright-ruby-client-1.37.1/lib/playwright/channel.rb:34:in `block in send_message_to_server_result'
2023-09-14T09:48:21.654992+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/playwright-ruby-client-1.37.1/lib/playwright/channel.rb:65:in `with_logging'
2023-09-14T09:48:21.654992+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/playwright-ruby-client-1.37.1/lib/playwright/channel.rb:33:in `send_message_to_server_result'
2023-09-14T09:48:21.654992+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/playwright-ruby-client-1.37.1/lib/playwright/channel.rb:20:in `send_message_to_server'
2023-09-14T09:48:21.654992+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/playwright-ruby-client-1.37.1/lib/playwright/channel_owners/browser_type.rb:14:in `launch'
2023-09-14T09:48:21.654993+00:00 app[worker.1]: /app/vendor/bundle/ruby/3.2.0/gems/playwright-ruby-client-1.37.1/lib/playwright_api/browser_type.rb:109:in `launch'
2023-09-14T09:48:21.654993+00:00 app[worker.1]: /app/app/models/scrappers/job.rb:7:in `block in getJobUrls'
```

Any recommendation on how to solve this issue? 

### Environment
- Rails 7
- Playright Gem 1.37.1
- Heroku

**Comment by cswilliams** [comment]
We're also seeing a similar issue.  Here's a stack trace from one of our freezes: 
```
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:785:in `sleep'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:785:in `wait'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:785:in `block in wait_until_resolved'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:773:in `synchronize'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:773:in `wait_until_resolved'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:1267:in `wait_until_resolved!'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:1482:in `value!'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.37.0/lib/playwright/connection.rb:104:in `send_message_to_server'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.37.0/lib/playwright/channel.rb:34:in `block in send_message_to_server_result'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.37.0/lib/playwright/channel.rb:65:in `with_logging'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.37.0/lib/playwright/channel.rb:33:in `send_message_to_server_result'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.37.0/lib/playwright/channel.rb:20:in `send_message_to_server'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.37.0/lib/playwright/channel_owners/frame.rb:177:in `wait_for_selector'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.37.0/lib/playwright/channel_owners/page.rb:233:in `wait_for_selector'
      /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.37.0/lib/playwright_api/page.rb:1611:in `wait_for_selector'
```

We have playwright deployed to an ECS Fargate service that runs a single playwright instance behind an ALB, where the ALB handles SSL termination.  When we redeploy the playwright service (ECS will spin up a new playwright instance, add it to the ALB pool, and then teardown the old playwright), I would say 50% of the time our code that uses the playwright client hangs with the above backtrace.  The other half of the time, we'll get a `SSL_write: protocol is shutdown` error (which is easy to rescue and retry from).

Will also try the latest version of the client once I get a chance.

**Comment by cswilliams** [comment]
I updated the gem to the latest version `1.41.1` and still see it hanging.  Here's a backtrace:
```
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:785:in `sleep'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:785:in `wait'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:785:in `block in wait_until_resolved'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:773:in `synchronize'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:773:in `wait_until_resolved'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:1267:in `wait_until_resolved!'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/concurrent-ruby-1.2.3/lib/concurrent-ruby/concurrent/promises.rb:1482:in `value!'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.41.1/lib/playwright/connection.rb:118:in `send_message_to_server'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.41.1/lib/playwright/channel.rb:35:in `block in send_message_to_server_result'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.41.1/lib/playwright/channel.rb:67:in `with_logging'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.41.1/lib/playwright/channel.rb:34:in `send_message_to_server_result'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.41.1/lib/playwright/channel.rb:20:in `send_message_to_server'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.41.1/lib/playwright/channel_owners/page.rb:342:in `reload'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.41.1/lib/playwright_api/page.rb:1085:in `reload'
```

**Comment by YusukeIwaki** [comment]
@cswilliams Hi, very sorry for no response.
Could you execute the script with `DEBUG=1` and provide a log around the error occurs?
I guess some event notifications from playwright driver are ignored/missed unexpectedly by playwright-ruby-client.

**Comment by cswilliams** [comment]
Sure, here you go:

```
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"remove"=>"commit"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"remove"=>"load"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"remove"=>"domcontentloaded"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"remove"=>"networkidle"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"add"=>"commit"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"navigated", "params"=>{"url"=>"https://XXXXXXXXXXX", "name"=>"", "newDocument"=>{"request"=>{"guid"=>"request@e9d03f1d7e1dbce06b223b81003031de"}}}}
RECV>{"guid"=>"browser-context@bbd47438250a28b9b159bca16f359ba1", "method"=>"requestFinished", "params"=>{"request"=>{"guid"=>"request@e9d03f1d7e1dbce06b223b81003031de"}, "response"=>{"guid"=>"response@4f79ce449ecfafb171e3936006a176f2"}, "responseEndTiming"=>3764.611, "page"=>{"guid"=>"page@11181cc114c6f06b410fa35c3236d4b5"}}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"add"=>"domcontentloaded"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"add"=>"load"}}
RECV>{"id"=>46, "result"=>{"response"=>{"guid"=>"response@4f79ce449ecfafb171e3936006a176f2"}}}
SEND>{:id=>47, :guid=>"frame@6968aeafe16d223ff32e33beb320c3fa", :method=>"waitForSelector", :params=>{:selector=>".link-record", :timeout=>5000}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"add"=>"networkidle"}}
RECV>{"id"=>47, "error"=>{"error"=>{"message"=>"Timeout 5000ms exceeded.\n=========================== logs ===========================\nwaiting for locator('.link-record') to be visible\n============================================================", "stack"=>"TimeoutError: Timeout 5000ms exceeded.\n=========================== logs ===========================\nwaiting for locator('.link-record') to be visible\n============================================================\n    at ProgressController.run (/home/pwuser/.npm/_npx/b8f8556362a868f7/node_modules/playwright-core/lib/server/progress.js:88:26)\n    at Frame.waitForSelector (/home/pwuser/.npm/_npx/b8f8556362a868f7/node_modules/playwright-core/lib/server/frames.js:618:23)\n    at FrameDispatcher.waitForSelector (/home/pwuser/.npm/_npx/b8f8556362a868f7/node_modules/playwright-core/lib/server/dispatchers/frameDispatcher.js:103:103)\n    at DispatcherConnection.dispatch (/home/pwuser/.npm/_npx/b8f8556362a868f7/node_modules/playwright-core/lib/server/dispatchers/dispatcher.js:313:46)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)", "name"=>"TimeoutError"}}}
SEND>{:id=>48, :guid=>"frame@6968aeafe16d223ff32e33beb320c3fa", :method=>"querySelector", :params=>{:selector=>".no-results-msg"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"__create__", "params"=>{"type"=>"ElementHandle", "initializer"=>{"preview"=>"JSHandle@node"}, "guid"=>"handle@acda47679b32e6b99b6dd3208c3f36e9"}}
RECV>{"id"=>48, "result"=>{"element"=>{"guid"=>"handle@acda47679b32e6b99b6dd3208c3f36e9"}}}
RECV>{"guid"=>"handle@acda47679b32e6b99b6dd3208c3f36e9", "method"=>"previewUpdated", "params"=>{"preview"=>"JSHandle@<span class=\"no-results-msg\">No records are available in this view</span>"}}
SEND>{:id=>49, :guid=>"page@11181cc114c6f06b410fa35c3236d4b5", :method=>"reload", :params=>{:timeout=>120000}}
RECV>{"guid"=>"page@11181cc114c6f06b410fa35c3236d4b5", "method"=>"__create__", "params"=>{"type"=>"Request", "initializer"=>{"frame"=>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa"}, "url"=>"https://XXXXXXXXXXX", "resourceType"=>"document", "method"=>"GET", "headers"=>[{"name"=>"Upgrade-Insecure-Requests", "value"=>"1"}, {"name"=>"User-Agent", "value"=>"Mozilla/5.0 (X11; Linux aarch64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/116.0.5845.82 Safari/537.36"}, {"name"=>"sec-ch-ua", "value"=>"\"Chromium\";v=\"116\", \"Not)A;Brand\";v=\"24\", \"HeadlessChrome\";v=\"116\""}, {"name"=>"sec-ch-ua-mobile", "value"=>"?0"}, {"name"=>"sec-ch-ua-platform", "value"=>"\"Linux\""}], "isNavigationRequest"=>true}, "guid"=>"request@d4a96cd8fd5814afb6e6bba48cd1b424"}}
RECV>{"guid"=>"browser-context@bbd47438250a28b9b159bca16f359ba1", "method"=>"request", "params"=>{"request"=>{"guid"=>"request@d4a96cd8fd5814afb6e6bba48cd1b424"}, "page"=>{"guid"=>"page@11181cc114c6f06b410fa35c3236d4b5"}}}
RECV>{"guid"=>"request@d4a96cd8fd5814afb6e6bba48cd1b424", "method"=>"__create__", "params"=>{"type"=>"Response", "initializer"=>{"request"=>{"guid"=>"request@d4a96cd8fd5814afb6e6bba48cd1b424"}, "url"=>"https://XXXXXXXXXXX", "status"=>200, "statusText"=>"", "headers"=>[{"name"=>"cache-control", "value"=>"private"}, {"name"=>"content-length", "value"=>"34664"}, {"name"=>"content-type", "value"=>"text/html; charset=utf-8"}, {"name"=>"date", "value"=>"Thu, 07 Mar 2024 00:11:35 GMT"}, {"name"=>"server", "value"=>"Microsoft-IIS/10.0"}, {"name"=>"x-aspnet-version", "value"=>"4.0.30319"}, {"name"=>"x-content-type-options", "value"=>"nosniff"}, {"name"=>"x-powered-by", "value"=>"ASP.NET"}, {"name"=>"x-xss-protection", "value"=>"1; mode=block"}], "timing"=>{"startTime"=>1709770292844.012, "domainLookupStart"=>-1, "domainLookupEnd"=>-1, "connectStart"=>-1, "secureConnectionStart"=>-1, "connectEnd"=>-1, "requestStart"=>0.455, "responseStart"=>2836.57}, "fromServiceWorker"=>false}, "guid"=>"response@df0fff516aa45e6b8a8c6b5d1fdbdcd2"}}
RECV>{"guid"=>"browser-context@bbd47438250a28b9b159bca16f359ba1", "method"=>"response", "params"=>{"response"=>{"guid"=>"response@df0fff516aa45e6b8a8c6b5d1fdbdcd2"}, "page"=>{"guid"=>"page@11181cc114c6f06b410fa35c3236d4b5"}}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"remove"=>"commit"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"remove"=>"domcontentloaded"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"remove"=>"load"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"remove"=>"networkidle"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"add"=>"commit"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"navigated", "params"=>{"url"=>"https://XXXXXXXXXXX", "name"=>"", "newDocument"=>{"request"=>{"guid"=>"request@d4a96cd8fd5814afb6e6bba48cd1b424"}}}}
RECV>{"guid"=>"browser-context@bbd47438250a28b9b159bca16f359ba1", "method"=>"requestFinished", "params"=>{"request"=>{"guid"=>"request@d4a96cd8fd5814afb6e6bba48cd1b424"}, "response"=>{"guid"=>"response@df0fff516aa45e6b8a8c6b5d1fdbdcd2"}, "responseEndTiming"=>2838.363, "page"=>{"guid"=>"page@11181cc114c6f06b410fa35c3236d4b5"}}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"add"=>"domcontentloaded"}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"add"=>"load"}}
RECV>{"id"=>49, "result"=>{"response"=>{"guid"=>"response@df0fff516aa45e6b8a8c6b5d1fdbdcd2"}}}
SEND>{:id=>50, :guid=>"frame@6968aeafe16d223ff32e33beb320c3fa", :method=>"waitForSelector", :params=>{:selector=>".link-record", :timeout=>5000}}
RECV>{"guid"=>"frame@6968aeafe16d223ff32e33beb320c3fa", "method"=>"loadstate", "params"=>{"add"=>"networkidle"}}
```

Just to make sure everything is clear, the hanging issue appears when the playwright server behind our load balancer is replaced. Our script that utilizes the playwright client connects to a load balancer over SSL, which then forwards the web socket connection to the underlying playwright server.  When this underlying playwright server is shutdown and a replacement is added, it seems like the client doesn't disconnect properly from the old server sometimes.  50% of the time our script exits with an `OpenSSL::SSL::SSLError: SSL_write: protocol is shutdown` (which is fine behavior and can be easily caught/retried), but the other half of the time it gets stuck in that promise and hangs forever.

**Comment by cswilliams** [comment]
Also, if it helps, this is what the backtrace looks like when it disconnects properly and doesn't hang: 
```
Error: OpenSSL::SSL::SSLError::SSL_write: protocol is shutdown
Backtrace: /Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/3.1.0/openssl/buffering.rb:354:in `syswrite'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/3.1.0/openssl/buffering.rb:354:in `do_write'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/3.1.0/openssl/buffering.rb:371:in `block in write'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/3.1.0/openssl/buffering.rb:370:in `each'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/3.1.0/openssl/buffering.rb:370:in `inject'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/3.1.0/openssl/buffering.rb:370:in `write'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.42.1/lib/playwright/web_socket_client.rb:42:in `write'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/websocket-driver-0.7.6/lib/websocket/driver/hybi.rb:229:in `send_frame'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/websocket-driver-0.7.6/lib/websocket/driver/hybi.rb:187:in `frame'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/websocket-driver-0.7.6/lib/websocket/driver/hybi.rb:151:in `close'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.42.1/lib/playwright/web_socket_client.rb:121:in `close'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.42.1/lib/playwright/web_socket_transport.rb:45:in `stop'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.42.1/lib/playwright/connection.rb:44:in `stop'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.42.1/lib/playwright.rb:48:in `stop'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.42.1/lib/playwright.rb:134:in `ensure in connect_to_playwright_server'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/playwright-ruby-client-1.42.1/lib/playwright.rb:134:in `connect_to_playwright_server'
```

**Comment by cswilliams** [comment]
If it helps, I made a branch with the fixes (based of 1.37.0 since that's what we currently use, I think it should apply cleanly to main tho): https://github.com/YusukeIwaki/playwright-ruby-client/compare/main...cswilliams:1.37.0_fix_hang?expand=1

Hopefully that's a bit easier to see.

Another issue I'm running into is that if I try to source this gem from either your github repo or my fork, like so:
Gemfile:
```
gem "playwright-ruby-client", git: 'https://github.com/YusukeIwaki/playwright-ruby-client'
```

Then run `bundle install` and try to run my script.  When calling `Playwright.connect_to_playwright_server` I get this error:
```
NoMethodError undefined method `new' for Playwright:Module

          @impl._api ||= api_class.new(@impl)
                                  ^^^^
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/bundler/gems/playwright-ruby-client-aeb9671a1b13/lib/playwright/playwright_api.rb:45:in `wrap'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/bundler/gems/playwright-ruby-client-aeb9671a1b13/lib/playwright/playwright_api.rb:13:in `wrap'
/Users/cswilliams/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/bundler/gems/playwright-ruby-client-aeb9671a1b13/lib/playwright.rb:124:in `connect_to_playwright_server'
```

Do you know why that might be? It only seems to appear when sourcing the gem from git in the Gemfile.

**Comment by cswilliams** [comment]
Please ignore my last question, I was finally able to get it working in a Gemfile via git after running `bundle exec ruby development/generate_api.rb`, adjusting the `.gitignore` files, and then committing the changed files into my branch.  Now that I am able to build my own gem with the fixes, I will do some more testing against my changes next week and let you know if I run into any issues.

**Comment by YusukeIwaki** [comment]
@cswilliams Thank you for your detailed information.

I tried the code and merged it into 1.42.x and 1.37.1.
Could you try playwright-ruby-client `1.37.2.alpha1` ? If it works well, I will set the version to 1.37.2.
https://github.com/YusukeIwaki/playwright-ruby-client/pull/292/files

**Comment by cswilliams** [comment]
@YusukeIwaki Unfortunately my contract just ended with the company that ran into this problem so I am not able to test it further.  However, I did build a custom gem from the changes I outlined above and deployed it to production several weeks ago.  We did not observe any more hangs (nor any other negative behavior) so I think it should be safe to use.
